Достигнута точность 99.2%

asemendarov opened the pull request:
Достигнута точность в **99.2%** для классов `'0, 1, 4, 7, 9'` из `data10mov_no_abs.mat`. Доля тренировочных данных к тестовым - 2 к 1 (задан параметром `test_size`).

### main_search_cv_1.py

Здесь реализована оценка качества обучения при четко заданных параметрах. Исключение составляют только параметры регуляризации SVM (параметры **C** и **gamma**): они подбираются при помощи [sklearn.model_selection.RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV).
 
### main_multiprocessing_3_and_search_cv_1.py

Здесь реализован поиск классов, которые дают максимальную точность. Метод поиска - полный перебор. Стоит отметить, что полный перебор осуществляется при жестко заданной длине комбинации классов (параметр `combinations_len`). Также для каждой комбинации осуществляется поиск параметр регуляризации SVM как и в `main_search_cv_1.py`.

Работу активно развиваю в своем репозитории - [emg_masterswork](https://github.com/oNoComments/emg_masterswork)

ps: ~~прошу прощения за не информативные названия файлов~~

asemendarov committed:
Added emg_platform/source/python/sav_work

asemendarov committed:
Update emg_platform/source/python/readme.md

estel1 commented:
Ок, спасибо. Я так понимаю что 99% - это на тестовых данных. В этом случае
очень хороший результат на ограниченном наборе. Скорее всего будут те же
трудности с новыми измерениями на которые указывал Андрей. Надо собирать
новую базу - побольше.


вт, 23 апр. 2019 г. в 12:01, oNoComments <notifications@github.com>:

> Достигнута точность в *99.2%* для классов '0, 1, 4, 7, 9' из
> data10mov_no_abs.mat. Доля тренировочных данных к тестовым - 2 к 1 (задан
> параметром test_size).
> main_search_cv_1.py
>
> Здесь реализована оценка качества обучения при четко заданных параметрах.
> Исключение составляют только параметры регуляризации SVM (параметры *C* и
> *gamma*): они подбираются при помощи
> sklearn.model_selection.RandomizedSearchCV
> <https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV>
> .
> main_multiprocessing_3_and_search_cv_1.py
>
> Здесь реализован поиск классов, которые дают максимальную точность. Метод
> поиска - полный перебор. Стоит отметить, что полный перебор осуществляется
> при жестко заданной длине комбинации классов (параметр combinations_len).
> Также для каждой комбинации осуществляется поиск параметр регуляризации SVM
> как и в main_search_cv_1.py.
>
> Работу активно развиваю в своем репозитории - emg_masterswork
> <https://github.com/oNoComments/emg_masterswork>
>
> ps: прошу прощения за не информативные названия файлов
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/RF-Lab/emg_platform/pull/2
> Commit Summary
>
>    - Added emg_platform/source/python/sav_work
>    - Update emg_platform/source/python/readme.md
>
> File Changes
>
>    - *M* source/python/readme.md
>    <https://github.com/RF-Lab/emg_platform/pull/2/files#diff-0> (21)
>    - *A* source/python/sav_work/main_multiprocessing_3_and_search_cv_1.py
>    <https://github.com/RF-Lab/emg_platform/pull/2/files#diff-1> (137)
>    - *A* source/python/sav_work/main_search_cv_1.py
>    <https://github.com/RF-Lab/emg_platform/pull/2/files#diff-2> (113)
>
> Patch Links:
>
>    - https://github.com/RF-Lab/emg_platform/pull/2.patch
>    - https://github.com/RF-Lab/emg_platform/pull/2.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RF-Lab/emg_platform/pull/2>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAZZWLCXBDYFCADXYCYDVTPR3F5JANCNFSM4HHV6M7A>
> .
>

venik commented on source/python/readme.md:
```diff
@@ -0,0 +1,21 @@
+Также для каждой комбинации осуществляется поиск параметр регуляризации SVM как и в `main_search_cv_1.py`.
+
+
+ps: ~~прошу прощения за не информативные названия файлов~~
```
useless comment

asemendarov replied:
Согласен. Будет исправлено.

venik commented on source/python/sav_work/main_multiprocessing_3_and_search_cv_1.py:
```diff
@@ -0,0 +1,137 @@
+
+    if return_result:
+        return result
+
```
useless

asemendarov replied:
Согласен. В своем репозитории  уже внес правки. Скоро это появится и здесь.

venik commented on source/python/sav_work/main_search_cv_1.py:
```diff
@@ -0,0 +1,113 @@
+            continue
+
+        X.extend(value)
+        y.extend([index] * len(value))
```
that would be extremely slow, you better use pandas slice or smth similar

asemendarov replied:
Могу предложить только такой вариант:
```
    X, y = [], []
    for index, value in enumerate(emg_data_set.iloc[classes_list]):
        X.extend(value)
        y.extend([index] * len(value))
```
Не вижу альтернативных решение для строк 89 и 90 (`X.extend(value)` и `y.extend([index] * len(value))`) так, чтобы не пострадала читабельность и/или эффективность. 

Я провел несколько тестов, которые должны убедить, что мой текущий код является вполне нормальным. В тесте я использовал ключевые операции, которые теоретически в любом случаи придется выполнить, если идти по пути поиска альтернативных решений. 

Код скрипта:
```
import timeit
import numpy as np

X = []
y = []

def test1():    # 1
    y = np.zeros(10)

def test2():    # 1
    y = [0] * 10

def test3():    # 2
    X = np.vstack([[11, 22, 33], [55, 66, 77]])

def test4():    # 2
    for value in [[11, 22, 33], [55, 66, 77]]:
        X.extend(value)

def test5():    # 3
    y = np.ravel([[0, 0, 0, 0], [1, 1, 1, 1]])

def test6():    # 3
    for value in [[0, 0, 0, 0], [1, 1, 1, 1]]:
        y.extend(value)



print(f'#1 test1(): {timeit.timeit("test1()", setup="from __main__ import test1", number=10000)}')
print(f'#1 test2(): {timeit.timeit("test2()", setup="from __main__ import test2", number=10000)}\n')
print(f'#2 test3(): {timeit.timeit("test3()", setup="from __main__ import test3", number=10000)}')
print(f'#2 test4(): {timeit.timeit("test4()", setup="from __main__ import test4", number=10000)}\n')
print(f'#3 test5(): {timeit.timeit("test5()", setup="from __main__ import test5", number=10000)}')
print(f'#3 test6(): {timeit.timeit("test6()", setup="from __main__ import test6", number=10000)}')
```

Результат выполнения:
```
#1 test1(): 0.008073533999999993
#1 test2(): 0.002052261999999999

#2 test3(): 0.11155472599999997
#2 test4(): 0.004853663000000008

#3 test5(): 0.03923518300000001
#3 test6(): 0.004528295999999987
```

venik replied:
нужно смотреть как данные выглядят. если фрейм будет гигабайт ^^^ все умрет и никогда больше не отомрет. Так никто не пишет.

Я не думаю что тест валиден, потому что данные одни и те же. Они кешируются.

asemendarov replied:
Было:
```
X, y = [], []
for index, value in enumerate(emg_data_set.iloc[classes_list]):
    X.extend(value)
    y.extend([index] * len(value))
```
Стало:
```
def _X(data):
    return np.vstack(data)

def _y(classes, data_len):
    return np.ravel(np.array(classes * data_len).reshape((data_len, -1)).T)

emg_data_set = emg_data_set.iloc[classes_list]
X, y = _X(emg_data_set), _y(classes_list, len(emg_data_set[0]))
```

venik commented on source/python/sav_work/main_multiprocessing_3_and_search_cv_1.py:
```diff
@@ -0,0 +1,137 @@
+# classes = '0, 1, 4, 6, 7'
+classes = '0, 1, 2, 3, 4, 5, 6, 7, 8, 9'
+
+svc = svm.SVC(kernel='rbf')
```
absolutely NO explanation why RBF was used, there is no comparison to anything else

asemendarov replied:
В своем репозитории я уже это сделал. Скоро появится и здесь.

venik commented on source/python/readme.md:
```diff
@@ -0,0 +1,21 @@
+## sav_work
+_SAV - Semendarov Andrey V._
+
+Достигнута точность в **99.2%** для классов `'0, 1, 4, 7, 9'` из `data10mov_no_abs.mat`. Доля тренировочных данных к тестовым - 2 к 1 (задан параметром `test_size`).
```
it requires more details on the method: the kernel, the optimal parameters, etc

asemendarov replied:
Хорошо. Здесь мне придется потратить некоторое время, чтобы описать все в необходимом качестве.

asemendarov committed:
Updated sav_feature

1. Удалил старую реализацию.
2. Обновил README.md (emg_platform\source\python\readme.md)
3. Добавил новую реализацию.
4. Добавил README.md (emg_platform\source\python\sav_work\readme.md), где отразил результаты исследования.
